File: frontend/src/components/benchmark/content/girvan_newman/implementations/girvan_newman.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class GirvanNewman
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    private class Edge
    {
        public int U { get; set; }
        public int V { get; set; }

        public Edge(int u, int v)
        {
            U = Math.Min(u, v);
            V = Math.Max(u, v);
        }

        public override bool Equals(object obj)
        {
            if (obj is Edge other)
            {
                return U == other.U && V == other.V;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(U, V);
        }
    }

    private Dictionary<Edge, double> edgeBetweenness = new Dictionary<Edge, double>();
    private List<List<int>> communities = new List<List<int>>();

    private void BfsDistances(int start, out Dictionary<int, int> dist, 
                              out Dictionary<int, List<int>> predecessors)
    {
        dist = new Dictionary<int, int>();
        predecessors = new Dictionary<int, List<int>>();
        var queue = new Queue<int>();

        dist[start] = 0;
        queue.Enqueue(start);

        while (queue.Count > 0)
        {
            int node = queue.Dequeue();

            foreach (int neighbor in adjacency[node])
            {
                if (!dist.ContainsKey(neighbor))
                {
                    dist[neighbor] = dist[node] + 1;
                    queue.Enqueue(neighbor);
                }
                if (dist[neighbor] == dist[node] + 1)
                {
                    if (!predecessors.ContainsKey(neighbor))
                        predecessors[neighbor] = new List<int>();
                    predecessors[neighbor].Add(node);
                }
            }
        }
    }

    private void ComputeEdgeBetweenness()
    {
        edgeBetweenness.Clear();
        foreach (var pair in adjacency)
        {
            foreach (int neighbor in pair.Value)
            {
                var edge = new Edge(pair.Key, neighbor);
                if (!edgeBetweenness.ContainsKey(edge))
                    edgeBetweenness[edge] = 0.0;
            }
        }

        foreach (int source in nodes)
        {
            BfsDistances(source, out var dist, out var predecessors);

            var dependency = new Dictionary<int, double>();
            foreach (int node in nodes)
                dependency[node] = 0.0;

            var sortedNodes = nodes.OrderByDescending(n => dist[n]).ToList();

            foreach (int node in sortedNodes)
            {
                if (!predecessors.ContainsKey(node)) continue;

                foreach (int pred in predecessors[node])
                {
                    double contrib = (1.0 + dependency[node]) / predecessors[node].Count;
                    dependency[pred] += contrib;
                }
            }

            foreach (int node in nodes)
            {
                if (!predecessors.ContainsKey(node)) continue;

                foreach (int pred in predecessors[node])
                {
                    var edge = new Edge(pred, node);
                    edgeBetweenness[edge] += dependency[node] / predecessors[node].Count;
                }
            }
        }

        var keys = edgeBetweenness.Keys.ToList();
        foreach (var key in keys)
        {
            edgeBetweenness[key] /= 2.0;
        }
    }

    private void RemoveEdgeWithMaxBetweenness()
    {
        var maxEdge = edgeBetweenness.OrderByDescending(x => x.Value).First().Key;

        adjacency[maxEdge.U].Remove(maxEdge.V);
        adjacency[maxEdge.V].Remove(maxEdge.U);
    }

    private List<List<int>> FindComponents()
    {
        var components = new List<List<int>>();
        var visited = new HashSet<int>();

        foreach (int node in nodes)
        {
            if (visited.Contains(node)) continue;

            var component = new List<int>();
            var queue = new Queue<int>();
            queue.Enqueue(node);
            visited.Add(node);

            while (queue.Count > 0)
            {
                int current = queue.Dequeue();
                component.Add(current);

                foreach (int neighbor in adjacency[current])
                {
                    if (!visited.Contains(neighbor))
                    {
                        visited.Add(neighbor);
                        queue.Enqueue(neighbor);
                    }
                }
            }

            if (component.Count > 0)
                components.Add(component);
        }

        return components;
    }

    private double Modularity(List<List<int>> communities)
    {
        var communityMap = new Dictionary<int, int>();
        for (int i = 0; i < communities.Count; i++)
        {
            foreach (int node in communities[i])
                communityMap[node] = i;
        }

        double m = 0.0;
        foreach (var pair in adjacency)
            m += pair.Value.Count;
        m /= 2.0;

        var degrees = new Dictionary<int, double>();
        foreach (int node in nodes)
            degrees[node] = adjacency[node].Count;

        double Q = 0.0;
        foreach (var pair in adjacency)
        {
            int u = pair.Key;
            foreach (int v in pair.Value)
            {
                if (communityMap[u] == communityMap[v])
                {
                    Q += 1.0 - (degrees[u] * degrees[v]) / (2.0 * m);
                }
            }
        }

        return Q / (2.0 * m);
    }

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u)) adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v)) adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public List<List<int>> DetectCommunities(int numCommunities = 2)
    {
        while (true)
        {
            ComputeEdgeBetweenness();
            RemoveEdgeWithMaxBetweenness();

            var currentComponents = FindComponents();
            if (currentComponents.Count >= numCommunities)
            {
                communities = currentComponents;
                break;
            }

            if (adjacency.Count == 0)
                break;
        }

        return communities;
    }

    public List<List<int>> DetectCommunitiesByModularity()
    {
        List<List<int>> bestCommunities = new List<List<int>>();
        double bestModularity = -1.0;
        int iterations = 0;

        while (true)
        {
            ComputeEdgeBetweenness();
            RemoveEdgeWithMaxBetweenness();

            var currentComponents = FindComponents();
            double currentModularity = Modularity(currentComponents);

            if (currentModularity > bestModularity)
            {
                bestModularity = currentModularity;
                bestCommunities = currentComponents;
            }

            if (currentComponents.Count == 1)
                break;

            iterations++;
            if (iterations > 1000) break;
        }

        return bestCommunities;
    }
}

File: frontend/src/components/benchmark/content/girvan_newman/implementations/girvan_newman.java
import java.util.*;

public class GirvanNewman {
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    private static class Edge {
        int u, v;

        Edge(int u, int v) {
            this.u = Math.min(u, v);
            this.v = Math.max(u, v);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Edge)) return false;
            Edge edge = (Edge) o;
            return u == edge.u && v == edge.v;
        }

        @Override
        public int hashCode() {
            return Objects.hash(u, v);
        }
    }

    private Map<Edge, Double> edgeBetweenness = new HashMap<>();
    private List<List<Integer>> communities = new ArrayList<>();

    private void bfsDistances(int start, Map<Integer, Integer> dist,
                              Map<Integer, List<Integer>> predecessors) {
        Queue<Integer> queue = new LinkedList<>();
        dist.put(start, 0);
        queue.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();

            for (int neighbor : adjacency.get(node)) {
                if (!dist.containsKey(neighbor)) {
                    dist.put(neighbor, dist.get(node) + 1);
                    queue.add(neighbor);
                }
                if (dist.get(neighbor) == dist.get(node) + 1) {
                    predecessors.putIfAbsent(neighbor, new ArrayList<>());
                    predecessors.get(neighbor).add(node);
                }
            }
        }
    }

    private void computeEdgeBetweenness() {
        edgeBetweenness.clear();

        for (Map.Entry<Integer, Set<Integer>> entry : adjacency.entrySet()) {
            int u = entry.getKey();
            for (int v : entry.getValue()) {
                Edge edge = new Edge(u, v);
                edgeBetweenness.putIfAbsent(edge, 0.0);
            }
        }

        for (int source : nodes) {
            Map<Integer, Integer> dist = new HashMap<>();
            Map<Integer, List<Integer>> predecessors = new HashMap<>();
            bfsDistances(source, dist, predecessors);

            Map<Integer, Double> dependency = new HashMap<>();
            for (int node : nodes) {
                dependency.put(node, 0.0);
            }

            List<Integer> sortedNodes = new ArrayList<>(dist.keySet());
            sortedNodes.sort((a, b) -> dist.get(b) - dist.get(a));

            for (int node : sortedNodes) {
                if (!predecessors.containsKey(node)) continue;

                for (int pred : predecessors.get(node)) {
                    double contrib = (1.0 + dependency.get(node)) / predecessors.get(node).size();
                    dependency.put(pred, dependency.get(pred) + contrib);
                }
            }

            for (int node : nodes) {
                if (!predecessors.containsKey(node)) continue;

                for (int pred : predecessors.get(node)) {
                    Edge edge = new Edge(pred, node);
                    edgeBetweenness.put(edge, edgeBetweenness.get(edge) +
                            dependency.get(node) / predecessors.get(node).size());
                }
            }
        }

        for (Map.Entry<Edge, Double> entry : edgeBetweenness.entrySet()) {
            entry.setValue(entry.getValue() / 2.0);
        }
    }

    private void removeEdgeWithMaxBetweenness() {
        Edge maxEdge = null;
        double maxBetweenness = -1.0;

        for (Map.Entry<Edge, Double> entry : edgeBetweenness.entrySet()) {
            if (entry.getValue() > maxBetweenness) {
                maxBetweenness = entry.getValue();
                maxEdge = entry.getKey();
            }
        }

        if (maxEdge != null) {
            adjacency.get(maxEdge.u).remove(maxEdge.v);
            adjacency.get(maxEdge.v).remove(maxEdge.u);
        }
    }

    private List<List<Integer>> findComponents() {
        List<List<Integer>> components = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();

        for (int node : nodes) {
            if (visited.contains(node)) continue;

            List<Integer> component = new ArrayList<>();
            Queue<Integer> queue = new LinkedList<>();
            queue.add(node);
            visited.add(node);

            while (!queue.isEmpty()) {
                int current = queue.poll();
                component.add(current);

                for (int neighbor : adjacency.get(current)) {
                    if (!visited.contains(neighbor)) {
                        visited.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }

            if (!component.isEmpty()) {
                components.add(component);
            }
        }

        return components;
    }

    private double modularity(List<List<Integer>> communities) {
        Map<Integer, Integer> communityMap = new HashMap<>();
        for (int i = 0; i < communities.size(); i++) {
            for (int node : communities.get(i)) {
                communityMap.put(node, i);
            }
        }

        double m = 0.0;
        for (Map.Entry<Integer, Set<Integer>> entry : adjacency.entrySet()) {
            m += entry.getValue().size();
        }
        m /= 2.0;

        Map<Integer, Double> degrees = new HashMap<>();
        for (int node : nodes) {
            degrees.put(node, (double) adjacency.get(node).size());
        }

        double Q = 0.0;
        for (Map.Entry<Integer, Set<Integer>> entry : adjacency.entrySet()) {
            int u = entry.getKey();
            for (int v : entry.getValue()) {
                if (communityMap.get(u).equals(communityMap.get(v))) {
                    Q += 1.0 - (degrees.get(u) * degrees.get(v)) / (2.0 * m);
                }
            }
        }

        return Q / (2.0 * m);
    }

    public void addEdge(int u, int v) {
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.putIfAbsent(v, new HashSet<>());
        adjacency.get(u).add(v);
        adjacency.get(v).add(u);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public List<List<Integer>> detectCommunities(int numCommunities) {
        while (true) {
            computeEdgeBetweenness();
            removeEdgeWithMaxBetweenness();

            List<List<Integer>> currentComponents = findComponents();
            if (currentComponents.size() >= numCommunities) {
                communities = currentComponents;
                break;
            }

            if (adjacency.isEmpty()) {
                break;
            }
        }

        return communities;
    }

    public List<List<Integer>> detectCommunitiesByModularity() {
        List<List<Integer>> bestCommunities = new ArrayList<>();
        double bestModularity = -1.0;
        int iterations = 0;

        while (true) {
            computeEdgeBetweenness();
            removeEdgeWithMaxBetweenness();

            List<List<Integer>> currentComponents = findComponents();
            double currentModularity = modularity(currentComponents);

            if (currentModularity > bestModularity) {
                bestModularity = currentModularity;
                bestCommunities = currentComponents;
            }

            if (currentComponents.size() == 1) {
                break;
            }

            iterations++;
            if (iterations > 1000) break;
        }

        return bestCommunities;
    }
}